crsqlite_extension/src/vtab.c

```diff
@@ -0,0 +1,462 @@
+  while( isspace((unsigned char)z[0]) ) z++;
+  return z;
+}
+
+/* Remove trailing whitespace from the end of string z[] */
+static void crsqlite_trim_whitespace(char *z){
+  size_t n = strlen(z);
+  while( n>0 && isspace((unsigned char)z[n]) ) n--;
+  z[n] = 0;
+}
+
+/* Dequote the string */
+static void crsqlite_dequote(char *z){
+  int j;
+  char cQuote = z[0];
+  size_t i, n;
+
+  if( cQuote!='\'' && cQuote!='"' ) return;
+  n = strlen(z);
+  if( n<2 || z[n-1]!=z[0] ) return;
+  for(i=1, j=0; i<n-1; i++){
+    if( z[i]==cQuote && z[i+1]==cQuote ) i++;
+    z[j++] = z[i];
+  }
+  z[j] = 0;
+}
+
+/* Check to see if the string is of the form:  "TAG = VALUE" with optional
+** whitespace before and around tokens.  If it is, return a pointer to the
+** first character of VALUE.  If it is not, return NULL.
+*/
+static const char *crsqlite_parameter(const char *zTag, int nTag, const char *z){
+  z = crsqlite_skip_whitespace(z);
+  if( strncmp(zTag, z, nTag)!=0 ) return 0;
+  z = crsqlite_skip_whitespace(z+nTag);
+  if( z[0]!='=' ) return 0;
+  return crsqlite_skip_whitespace(z+1);
+}
+
+#if 0 /* not used - yet */
+/* Return 0 if the argument is false and 1 if it is true.  Return -1 if
+** we cannot really tell.
+*/
+static int crsqlite_boolean(const char *z){
+  if( sqlite3_stricmp("yes",z)==0
+   || sqlite3_stricmp("on",z)==0
+   || sqlite3_stricmp("true",z)==0
+   || (z[0]=='1' && z[1]==0)
+  ){
+    return 1;
+  }
+  if( sqlite3_stricmp("no",z)==0
+   || sqlite3_stricmp("off",z)==0
+   || sqlite3_stricmp("false",z)==0
+   || (z[0]=='0' && z[1]==0)
+  ){
+    return 0;
+  }
+  return -1;
+}
+#endif
+
+/*
+** The crsqliteConnect() method is invoked to create a new
+** crsqlite_vtab that describes the crsqlite virtual table.
+**
+** Think of this routine as the constructor for crsqlite_vtab objects.
+**
+** All this routine needs to do is:
+**
+**    (1) Allocate the crsqlite_vtab object and initialize all fields.
+**
+**    (2) Tell SQLite (via the sqlite3_declare_vtab() interface) what the
+**        result set of queries against crsqlite will look like.
+*/
+static int crsqliteCreate(
+  sqlite3 *db,
+  void *pAux,
+  int argc, const char *const*argv,
+  sqlite3_vtab **ppVtab,
+  char **pzErr
+){
+  //TODO: implement crsqlite_vtab object
+  crsqlite_vtab *pNew;
+  int i;
+  int rc;
+
+  //Create comma seperated list of arguments
+  sqlite3_str* createTableArgs = sqlite3_str_new(NULL);
+  for(int i = 3; i < argc; i++){
+    sqlite3_str_appendall(createTableArgs, argv[i]);
+    if (i != argc -1){
+      sqlite3_str_appendall(createTableArgs, ", ");
+    }
+  }
+
+  char *createTableString = sqlite3_str_finish(createTableArgs);
+
+  rc = init_storage(db, argv[2], createTableString, pzErr);
+  if (rc!=SQLITE_OK){
+    *pzErr = "Initializing storage failed\n";
+    sqlite3_free(createTableString);
+    return rc;
+  }
+  
+  char* declareSql = sqlite3_mprintf("CREATE TABLE sqliteIgnoresThisName(%s);", createTableString);
```

Comment:
sqliteIgnoresThisName -- what name ends up getting assigned? The name the user provided in

`CREATE VIRTUAL TABLE table_name`?

Reply:
Thats correct, sql doesnt care about the table name -- just the columns.